README
NAME
    Email::Abstract - Unified interface to mail representations

SYNOPSIS
      my $message = Mail::Message->read($rfc822)
                    || Email::Simple->new($rfc822)
                    || Mail::Internet->new([split /\n/, $rfc822])
                    || ...;

      my $subject = Email::Abstract->get_header($message, "Subject");
      Email::Abstract->set_header($message, "Subject", "My new subject");

      my $body = Email::Abstract->get_body($message);
      Email::Abstract->set_body($message, "Hello\nTest message\n");

      $rfc822 = Email::Abstract->as_string($message);

      my $mail_message = Email::Abstract->cast($message, "Mail::Message");

DESCRIPTION
    "Email::Abstract" provides module writers with the ability to write
    representation-independent mail handling code. For instance, in the
    cases of "Mail::Thread" or "Mail::ListDetector", a key part of the code
    involves reading the headers from a mail object. Where previously one
    would either have to specify the mail class required, or to build a new
    object from scratch, "Email::Abstract" can be used to perform certain
    simple operations on an object regardless of its underlying
    representation.

    "Email::Abstract" currently supports "Mail::Internet", "MIME::Entity",
    "Mail::Message", "Email::Simple" and "Email::MIME". Other
    representations are encouraged to create their own "Email::Abstract::*"
    class by copying "Email::Abstract::EmailSimple". All modules installed
    under the "Email::Abstract" hierarchy will be automatically picked up
    and used.

METHODS
  get_header($obj, $header)
    This returns the value or list of values of the given header.

  set_header($obj, $header, @lines)
    This sets the $header header to the given one or more values.

  get_body($obj)
    This returns the body as a string.

  set_body($obj, $string)
    This changes the body of the email to the given string.

  as_string($obj)
    This returns the whole email as a string.

AUTHOR
    Casey West, <casey@geeknest.com>

    Simon Cozens, <simon@cpan.org>

COPYRIGHT AND LICENSE
    Copyright 2004 by Simon Cozens

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    http://emailproject.perl.org/wiki/Email::Abstract



Changes
Revision history for Email-Abstract.

{{$NEXT}}

3.010     2023-09-02 20:22:12-04:00 America/New_York
        - update packaging metadata

3.009     2021-06-20 16:54:40-04:00 America/New_York
        - MIME::Entity 5.508 is required for operation
        - update author information
        - add a perl support section to docs

3.008     2014-12-27 18:36:19-05:00 America/New_York
        - make results of get_body be the same on Email::{Simple,MIME}
        - ...but this method is a mess, so maybe avoid using Abstract for body
          work

3.007     2013-12-31 10:39:14 America/New_York
          fix skip count when MIME::Entity is not present

3.006     2013-09-01 08:41:48 America/New_York
          require Email::Simple 1.998 to allow ->new(\$string)

3.005     2013-07-30 09:57:36 America/New_York
          repackage, update bug tracker, etc.

          provide raw body on multipart Email::MIME messages (thanks,
          bailey@upenn)
          [rt.cpan.org #77624]

3.004     2011-02-18

          If present, MIME::Entity must be v5.501; v5.500 had a regression (or
          a bug fix, depending how you look at it) that broke header-reading.
          While technically older versions that are not 5.500 would work, it is
          much simpler to just require the newest version, rather than to
          support a version range with a hole in it.

3.003     2010-11-08
          replace Class::ISA with mro (MRO::Compat)

3.002     2010-06-11
          avoid a warning in MailInternet with zero headers found

3.001     2008-12-11
          declare our prereq on Scalar::Util

3.000     2008-12-09
          accept scalar references for more efficient msg-from-string

2.134     2007-11-16
          (no code changes from previous dev release)

2.133_05  2007-11-11
          [BUG FIXES]
          added is_available method to MIMEEntity plugin

2.133_04  2007-09-24
          [ENHANCEMENTS]
          created Email::Abstract::Plugin base class; please use it!
          added is_available method to plugins

          [BUG FIXES]
          is_available in the Mail::Internet adapter should solve header
            folding issues (by preventing you from using it when it can't work)

2.133_03  2007-08-??
          diagnostics in output to indicate what version of a module we used

2.133_02  2007-07-??
          fix test planning

2.133_01  2007-07-??
          add test to ensure that "can't handle" exception is thrown ASAP
          remove unexplained requirement for perl 5.6
          fix Mail::Internet header fetching to unfold headers
          fix Mail::Message body setter, which hosed newlines
          fix body handling for Mail::Internet
          improved consistency of method used to find adapter class
          improved tests and test coverage

2.132     2007-03-22
          packaging improvements

2.131     2006-08-22
          pod tests

2.13      2006-07-24
          test for and permit passing Email::Abstract objects to Email::Abstract
          class methods

2.12      2006-07-24
          don't use MIME::Entity in test if it's not available

2.11      2006-07-22
          better test planning

2.10      2006-07-21
          add a new method to create wrapper objects
          handle subclasses /properly/ (correct ISA order)
          improved tests and test coverage
          miscellaneous refactoring
          update PEP URL
          update documentation

2.01      2004-11-04
          Minor Documentation Fix
          Author Change
          PEP Contact Added

2.0       2004-08-25 12:12:37 BST
          Handle subclasses

0.01      2004-05-26 16:47 20
          original version; created by h2xs 1.22 with options
            -AX -b 5.6.0 -n Email::Abstract



lib/Email/Abstract.pm
use 5.006;
use warnings;
use strict;
package Email::Abstract;
# ABSTRACT: unified interface to mail representations

use Carp;
use Email::Simple;
use MRO::Compat;

use Module::Pluggable 1.5
  search_path => [__PACKAGE__],
  except      => 'Email::Abstract::Plugin',
  require     => 1;

use Scalar::Util ();

my @plugins = __PACKAGE__->plugins(); # Requires them.
my %adapter_for =
  map  { $_->target => $_ }
  grep {
    my $avail = eval { $_->is_available };
    $@ ? ($@ =~ /Can't locate object method "is_available"/) : $avail;
  }
  @plugins;

sub object {
  my ($self) = @_;
  return unless ref $self;
  return $self->[0];
}

sub new {
  my ($class, $foreign) = @_;

  return $foreign if eval { $foreign->isa($class) };

  $foreign = Email::Simple->new($foreign)
    unless Scalar::Util::blessed($foreign);

  my $adapter = $class->__class_for($foreign); # dies if none available
  return bless [ $foreign, $adapter ] => $class;
}

sub __class_for {
  my ($self, $foreign, $method, $skip_super) = @_;
  $method ||= 'handle';

  my $f_class = ref $foreign;
     $f_class = $foreign unless $f_class;

  return $f_class if ref $foreign and $f_class->isa($self);

  return $adapter_for{$f_class} if $adapter_for{$f_class};

  if (not $skip_super) {
    my @bases = @{ mro::get_linear_isa($f_class) };
    shift @bases;
    for my $base (@bases) {
      return $adapter_for{$base} if $adapter_for{$base};
    }
  }

  Carp::croak "Don't know how to $method $f_class";
}

sub _adapter_obj_and_args {
  my $self = shift;

  if (my $thing = $self->object) {
    return ($self->[1], $thing, @_);
  } else {
    my $thing   = shift;
    my $adapter = $self->__class_for(
      Scalar::Util::blessed($thing) ? $thing : 'Email::Simple'
    );
    return ($adapter, $thing, @_);
  }
}

for my $func (qw(get_header get_body set_header set_body as_string)) {
  no strict 'refs';
  *$func = sub {
    my $self = shift;
    my ($adapter, $thing, @args) = $self->_adapter_obj_and_args(@_);

    # In the event of Email::Abstract->get_body($email_abstract), convert
    # it into an object method call.
    $thing = $thing->object if eval { $thing->isa($self) };

    # I suppose we could work around this by leaving @_ intact and assigning to
    # it.  That seems ... not good. -- rjbs, 2007-07-18
    unless (Scalar::Util::blessed($thing)) {
      Carp::croak "can't alter string in place" if substr($func, 0, 3) eq 'set';
      $thing = Email::Simple->new(
        ref $thing ? \do{my$str=$$thing} : $thing
      );
    }

    return $adapter->$func($thing, @args);
  };
}

sub cast {
  my $self = shift;
  my ($from_adapter, $from, $to) = $self->_adapter_obj_and_args(@_);

  my $adapter = $self->__class_for($to, 'construct', 1);

  my $from_string = ref($from) ? $from_adapter->as_string($from) : $from;

  return $adapter->construct($from_string);
}

1;
__END__

=head1 SYNOPSIS

  my $message = Mail::Message->read($rfc822)
             || Email::Simple->new($rfc822)
             || Mail::Internet->new([split /\n/, $rfc822])
             || ...
             || $rfc822;

  my $email = Email::Abstract->new($message);

  my $subject = $email->get_header("Subject");
  $email->set_header(Subject => "My new subject");

  my $body = $email->get_body;

  $rfc822 = $email->as_string;

  my $mail_message = $email->cast("Mail::Message");

=head1 DESCRIPTION

C<Email::Abstract> provides module writers with the ability to write
simple, representation-independent mail handling code. For instance, in the
cases of C<Mail::Thread> or C<Mail::ListDetector>, a key part of the code
involves reading the headers from a mail object. Where previously one would
either have to specify the mail class required, or to build a new object from
scratch, C<Email::Abstract> can be used to perform certain simple operations on
an object regardless of its underlying representation.

C<Email::Abstract> currently supports C<Mail::Internet>, C<MIME::Entity>,
C<Mail::Message>, C<Email::Simple>, C<Email::MIME>, and C<Courriel>.  Other
representations are encouraged to create their own C<Email::Abstract::*> class
by copying C<Email::Abstract::EmailSimple>.  All modules installed under the
C<Email::Abstract> hierarchy will be automatically picked up and used.

=head1 METHODS

All of these methods may be called either as object methods or as class
methods.  When called as class methods, the email object (of any class
supported by Email::Abstract) must be prepended to the list of arguments, like
so:

  my $return = Email::Abstract->method($message, @args);

This is provided primarily for backwards compatibility.

=head2 new

  my $email = Email::Abstract->new($message);

Given a message, either as a string or as an object for which an adapter is
installed, this method will return a Email::Abstract object wrapping the
message.

If the message is given as a string, it will be used to construct an object,
which will then be wrapped.

=head2 get_header

  my $header  = $email->get_header($header_name);

  my @headers = $email->get_header($header_name);

This returns the values for the given header.  In scalar context, it returns
the first value.

=head2 set_header

  $email->set_header($header => @values);

This sets the C<$header> header to the given one or more values.

=head2 get_body

  my $body = $email->get_body;

This returns the body as a string.

=head2 set_body

  $email->set_body($string);

This changes the body of the email to the given string.

B<WARNING!>  You probably don't want to call this method, despite what you may
think.  Email message bodies are complicated, and rely on things like content
type, encoding, and various MIME requirements.  If you call C<set_body> on a
message more complicated than a single-part seven-bit plain-text message, you
are likely to break something.  If you need to do this sort of thing, you
should probably use a specific message class from end to end.

This method is left in place for backwards compatibility.

=head2 as_string

  my $string = $email->as_string;

This returns the whole email as a decoded string.

=head2 cast

  my $mime_entity = $email->cast('MIME::Entity');

This method will convert a message from one message class to another.  It will
throw an exception if no adapter for the target class is known, or if the
adapter does not provide a C<construct> method.

=head2 object

  my $message = $email->object;

This method returns the message object wrapped by Email::Abstract.  If called
as a class method, it returns false.

Note that, because strings are converted to message objects before wrapping,
this method will return an object when the Email::Abstract was constructed from
a string.

=cut
